server/src/test/java/org/elasticsearch/action/support/replication/PostWriteRefreshTests.java

```diff
@@ -0,0 +1,207 @@
+            );
+            when(routingTable.unpromotableShards()).thenReturn(List.of(shardRouting));
+            WriteRequest.RefreshPolicy policy = randomFrom(WriteRequest.RefreshPolicy.IMMEDIATE, WriteRequest.RefreshPolicy.WAIT_UNTIL);
+            postWriteRefresh.refreshShard(policy, primary, result.getTranslogLocation(), f);
+            final Releasable releasable;
+            if (policy == WriteRequest.RefreshPolicy.WAIT_UNTIL) {
+                releasable = simulatePeriodicRefresh(primary);
+            } else {
+                releasable = () -> {};
+            }
+            f.actionGet();
+            assertTrue(unpromtableRefreshRequestReceived.get());
+            assertEngineContainsIdNoRefresh(primary, id);
+            releasable.close();
+        } finally {
+            closeShards(primary, primary);
+        }
+    }
+
+    public void testWaitUntilRefreshReplicaShard() throws IOException {
+        final IndexShard primary = newShard(true);
+        recoverShardFromStore(primary);
+        final IndexShard replica = newShard(false);
+        recoverReplica(replica, primary, true);
+        try {
+            String id = "0";
+            Engine.IndexResult result = indexDoc(replica, "_doc", id);
+            PlainActionFuture<Boolean> f = PlainActionFuture.newFuture();
+            PostWriteRefresh.refreshReplicaShard(WriteRequest.RefreshPolicy.WAIT_UNTIL, replica, result.getTranslogLocation(), f);
+            Releasable releasable = simulatePeriodicRefresh(replica);
+            f.actionGet();
+            assertEngineContainsIdNoRefresh(replica, id);
+            releasable.close();
+        } finally {
+            closeShards(primary, replica);
+        }
+    }
+
+    public void testImmediateRefreshReplicaShard() throws IOException {
+        final IndexShard primary = newShard(true);
+        recoverShardFromStore(primary);
+        final IndexShard replica = newShard(false);
+        recoverReplica(replica, primary, true);
+        try {
+            String id = "0";
+            Engine.IndexResult result = indexDoc(replica, "_doc", id);
+            PlainActionFuture<Boolean> f = PlainActionFuture.newFuture();
+            PostWriteRefresh.refreshReplicaShard(WriteRequest.RefreshPolicy.IMMEDIATE, replica, result.getTranslogLocation(), f);
+            f.actionGet();
+            assertEngineContainsIdNoRefresh(replica, id);
+        } finally {
+            closeShards(primary, replica);
+        }
+    }
+
+    private static void assertEngineContainsIdNoRefresh(IndexShard replica, String id) throws IOException {
+        List<DocIdSeqNoAndSource> docIds = EngineTestCase.getDocIds(replica.getEngineOrNull(), false);
+        Set<String> ids = docIds.stream().map(DocIdSeqNoAndSource::id).collect(Collectors.toSet());
+        assertThat(ids, contains(id));
+    }
+
+    private static Releasable simulatePeriodicRefresh(IndexShard shard) {
+        // Simulate periodic refresh
```

Comment:
Why is this periodic? Seems like a one-shot. Maybe rename to `delayed`?

Reply:
Periodic refers to the type of refresh it is emulating. But I will probably rename it to scheduled refresh. See IndexShard#scheduledRefresh.